### PR TITLE
Master test for simulator and post-processing

### PR DESCRIPTION
**Pull request by unie31**
Closing #119 

Note that the file [tests/resources/two_vmi12_stands_as_jsonpickle.json](https://github.com/menu-hanke/sim-workbench/pull/124/files#diff-b04782069daa70a300b2b0eee424f7038583763299144ae185b3d747d5e6432a) used in the simulator main test has been created using this change in vmi-data-converter: https://github.com/menu-hanke/vmi-data-converter/pull/190

### CODE REVIEW

File: tests/sim/main_test.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+                    'input_file': input_file,
+                    'control_file': 'control.yaml',
+                    'output_file': 'tests/sim/temp_output.pickle',
+                    'strategy': strategy,
+                    'input_format': format
+                }
+
+                simulator.main(args)
+
+                self.assertTrue(True)
```
Comment (unie31) [inline tests/sim/main_test.py:28]: I assume this kind of "heuristic" assertion is enough, or should the simulator be made to return a value on successful completion?

Comment (dmela) [reply tests/sim/main_test.py:28]: I would probably tests if the file exists... i.e. if the runtime was able to produce the output.
But in this case it is more a matter of taste. I would say it is up to you.

File: tests/sim/main_test.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import os
+import unittest
+import app.simulator as simulator
+import app.post_processing as pp
+
+class MainTest(unittest.TestCase):
+    
+    def test_sim_main(self):
+        """Tests the main function of the simulator with different parameter combinations. 
+        NOTE: it uses the control.yaml in the root of the project to easily test the most recent version of it."""
```
Comment (dmela) [inline tests/sim/main_test.py:10]: I would move this and the postp_processing docstring as module docstring. They are so similar...

File: app/app_io.py [hunk 3]
```diff
@@ -1,6 +1,9 @@
 import argparse
 from typing import List
 
+def function_arguments_to_argparse_namespace(args: dict):
```
Comment (dmela) [inline app/app_io.py:4]: Missing signature output type. I assume it is argparse.Namespace?

Comment (dmela) [reply app/app_io.py:4]: or SimpleNamespace?

File: app/post_processing.py [hunk 4]
```diff
@@ -1,16 +1,22 @@
...
 from forestry.operations import operation_lookup
 from sim.core_types import OperationPayload
 from sim.generators import simple_processable_chain
 from sim.runners import evaluate_sequence
 
 
-def main():
-    app_arguments = post_processing_cli_arguments(sys.argv[1:])
+def main(test_automation_arguments: dict = None):
+    """:test_automation_arguments: are only passed in from a test that ensures the post processing main remains runnable after feature changes."""
```
Comment (dmela) [inline app/post_processing.py:13]: I would like to say this comment is not necessary. I think the variable name already holds enough detail for the reader to understand that `test_automation_arguments` are for testing.

I could imagine that eventually such a docstring will start to rot. Purpose of a docstring is to describe what the function does, so I would add a more detailed description about what is the purpose of the main function or then remove the whole line.

File: app/app_io.py (hunk 3 above)
Comment (dmela) [reply app/app_io.py:4]: Also would consider renaming the function to e.g. test_arguments etc.

File: app/post_processing.py [hunk 5]
```diff
@@ -1,16 +1,22 @@
...
+
+    if test_automation_arguments is not None:
+        app_arguments = function_arguments_to_argparse_namespace(test_automation_arguments)
+    else:
+        app_arguments = post_processing_cli_arguments(sys.argv[1:])
```
Comment (dmela) [inline app/post_processing.py:18]: Could we write this as a onelier?
```
app_arguments = post_processing_cli_arguments(sys.argv[1:]) if test_automaton_arguments is None else function_arguments_to_argparse_namespace(test_automation_arguments)
```

**Review by dmela — CHANGES_REQUESTED** [review]
Looking good. Lefts few improvement ideas for you to consider.

File: app/post_processing.py (hunk 5 above)
Comment (unie31) [reply app/post_processing.py:18]: I'd say that is less pleasant to read

**Comment by dmela** [comment]
Try to refactor the solution to use `sys.argv` so that the `app` library would not be dependent from the `tests` library. In order to do this drop the `test_arguments_to_argparse_namespace` and refactor the test suites so that `sys.argv` would be inflated before calling the `main` function. If this is possible you can use functions in `app.app_io` module to read the "CLI" arguments.

**Review by dmela — APPROVED** [review]
Great work!

**Pull request merged**